docs/Off-line-Signatures-QR-Code.md

```diff
@@ -192,6 +211,55 @@ Generate the QR code from `offlineData`. Code example in Java:
+
+#### 2.1. Verify ECDSA Signature of Offline Data
+
+After user scans the QR code using a mobile app, the `ECDSA_SIGNATURE` should be validated on mobile device to verify authenticity of received data by taking contents of `offlineData` before the `ECDSA_SIGNATURE` and computing the ECDSA signature using `KEY_SERVER_PRIVATE`. Both signatures must match before continuing with the offline data signature verification.
+
+#### 2.2. Computing the Signature
+
+Mobile device prompts the user for the PIN code or use of a biometry and computes 4x4 digit long authentication code to be rewritten manually.
+
+### 3. Validating the Signature
+
+#### 3.1. Processing The User Input
+
+After user enters 4x4 digits in the browser, the value must be converted into a standard PowerAuth signature format that uses 2x8 digits. For example, an offline signature `1234-5678-9012-3456` needs to be converted into `12345678-90123456` (removing 1st and 3rd dash).
+
+#### 3.2. Preparing Signature Base String
+
+Now, you need to prepare a normalized data package called "signature base string". This is the payload that mobile app used to compute the signature - you need to have the same signature base string in order to be able to verify the signature.
+
+To compute the signature base string, you need:
+
+- `data` (as obtained in 1.1.)
+- `nonce` value (as obtained in 1.2)
+- two static constants: `POST` and `/operation/authorize/offline`
+
+The [algorithm for signature data normalization](https://developers.wultra.com/docs/develop/powerauth-crypto/Computing-and-Validating-Signatures#normalized-data-for-http-requests) is available in the cryptography description. The Java code for computing the normalized signature base string follows:
+
+```java
+String signatureBaseString
+    = PowerAuthHttpBody.getSignatureBaseString(
+        "POST",
+        "/operation/authorize/offline",
+        BaseEncoding.base64().decode(nonce),
+        data.getBytes()
+);
+```
+
+#### 3.4. Verifying Signature
+
+To verify signature, you need to call the SOAP method [`verifyOfflineSignature`](https://developers.wultra.com/docs/develop/powerauth-server/SOAP-Service-Methods#method-verifyofflinesignature) providing:
+
+- `signatureBaseString` (as obtained in 3.2.) - as data for verification
+- `signature` - value of the signature entered by the user (as obtained in 3.1., 2x8 digits)
+- `activationId` - identifier of the activation (to know which device is responsible for verification)
+- `signatureType` - type of the signature (`POSSESSION_KNOWLEDGE`).
 
-The `ECDSA_SIGNATURE` should be validated on mobile device to verify authenticity of received data by taking contents of `offlineData` before the `ECDSA_SIGNATURE` and computing the ECDSA signature using `KEY_SERVER_PRIVATE`. Both signatures must match before continuing with the offline data signature verification.
+The method returns information about signature verification, see the SOAP method documentation for details.
```

Comment:
Perhaps add comment about how to check the result of verification?

Reply:
Done.